libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,564 @@
+		result.emplace_back(Operation{Location::Storage, Effect::Read, {}, {}});
+
+	return result;
+}
+
+void RedundantStoreEliminator::applyOperation(RedundantStoreEliminator::Operation const& _operation)
+{
+	for (auto& [statement, state]: m_stores)
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool RedundantStoreEliminator::knownUnrelated(
+	RedundantStoreEliminator::Operation const& _op1,
+	RedundantStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+	}
+	else
+	{
+		u256 largestPositive = (u256(1) << 128) - 1;
+
+		yulAssert(_op1.location == Location::Memory, "");
+		if (_op1.length && knowledge.knownToBeZero(*_op1.length))
+			return true;
+		if (_op2.length && knowledge.knownToBeZero(*_op2.length))
+			return true;
+		if (!_op1.start || !_op2.start || !_op1.length || !_op2.length)
+			// Can we say anything else here?
+			return false;
+
+		// 1.start + 1.length <= 2.start ||
+		if (optional<u256> length1 = knowledge.knownConstant(*_op1.length))
+			if (optional<u256> diff = knowledge.differenceIsKnownConstant(*_op2.start, *_op1.start))
+				// diff = 2.start - 1.start
+				if (*length1 <= *diff && *length1 <= largestPositive && *diff <= largestPositive)
+				{
+					//cout << "Knon unrelated: length1 " << *length1 << ", diff: " << *diff << endl;
+					return true;
+				}
+		// 2.start + 2.length <= 1.start
+		if (optional<u256> length2 = knowledge.knownConstant(*_op2.length))
+			if (optional<u256> diff = knowledge.differenceIsKnownConstant(*_op1.start, *_op2.start))
+				// diff = 1.start - 2.start
+				if (*length2 <= *diff && *length2 <= largestPositive && *diff <= largestPositive)
+				{
+					//cout << "Knon unrelated: length2 " << *length2 << ", diff: " << *diff << endl;
+					return true;
+				}
+	}
+
+	return false;
+}
+
+bool RedundantStoreEliminator::knownCovered(
+	RedundantStoreEliminator::Operation const& _covered,
+	RedundantStoreEliminator::Operation const& _covering
+) const
+{
+	if (_covered.location != _covering.location)
+		return false;
+	if (_covered.start && _covered.length && _covered.start == _covering.start && _covered.length == _covering.length)
+		return true;
+	if (_covered.location == Location::Storage && _covered.start && _covering.start && *_covered.start == *_covering.start)
+		return true;
+	if (_covered.location == Location::Memory)
+	{
+		KnowledgeBase knowledge(m_dialect, m_ssaValues);
+		u256 largestPositive = (u256(1) << 128) - 1;
```

Comment:
Why was this number chosen? Does this have to do with EIP-1985? The largest value of `msize` there is `2**32 - 1`.

Reply:
Please check the code, but this is for the check `i.start <= e.start && e.start + e.length <= i.start + i.length`, it's unrelated to msize or any assumptions about memory size.